lib/linter/linter.js

```diff
@@ -2057,77 +2088,78 @@ class Linter {
      * @returns {(LintMessage|SuppressedLintMessage)[]} The found problems.
      */
     _verifyWithProcessor(textOrSourceCode, config, options, configForRecursive) {
+        const slots = internalSlotsMap.get(this);
         const filename = options.filename || "<input>";
         const filenameToExpose = normalizeFilename(filename);
         const physicalFilename = options.physicalFilename || filenameToExpose;
         const text = ensureText(textOrSourceCode);
+        const file = new VFile(filenameToExpose, text, {
+            physicalPath: physicalFilename
+        });
+
         const preprocess = options.preprocess || (rawText => [rawText]);
         const postprocess = options.postprocess || (messagesList => messagesList.flat());
-        const filterCodeBlock =
-            options.filterCodeBlock ||
-            (blockFilename => blockFilename.endsWith(".js"));
-        const originalExtname = path.extname(filename);
 
-        let blocks;
-
-        try {
-            blocks = preprocess(text, filenameToExpose);
-        } catch (ex) {
+        const processorService = new ProcessorService();
+        const preprocessResult = processorService.preprocessSync(file, {
+            processor: {
+                preprocess,
+                postprocess
+            }
+        });
 
-            // If the message includes a leading line number, strip it:
-            const message = `Preprocessing error: ${ex.message.replace(/^line \d+:/iu, "").trim()}`;
+        if (!preprocessResult.ok) {
+            return preprocessResult.errors;
+        }
 
-            debug("%s\n%s", message, ex.stack);
+        const filterCodeBlock =
+            options.filterCodeBlock ||
+            (blockFilePath => blockFilePath.endsWith(".js"));
+        const originalExtname = path.extname(filename);
 
-            return [
-                {
-                    ruleId: null,
-                    fatal: true,
-                    severity: 2,
-                    message,
-                    line: ex.lineNumber,
-                    column: ex.column,
-                    nodeType: null
-                }
-            ];
-        }
+        const { files } = preprocessResult;
 
-        const messageLists = blocks.map((block, i) => {
-            debug("A code block was found: %o", block.filename || "(unnamed)");
+        const messageLists = files.map(block => {
+            debug("A code block was found: %o", block.path ?? "(unnamed)");
```

Comment:
```suggestion
            debug("A code block was found: %o", block.path);
```

Same as https://github.com/eslint/eslint/pull/18818#discussion_r1764787354, though I think it would be better to move both into `ProcessorService#preprocessSync()`.